### ISSUE TITLE
Unclear issue with incremental compile

### ISSUE THREAD

**Issue by SwampWalker** [rust-lang/rust#92817]
### Code

Please contact me and I will give you access to my private repository on github if you are interesting in reproduction. The code isn't enormous, but it's unclear what is causing it. I also have some workarounds available so it is not urgent (this is a shallow justification for me not trying to prepare a smaller reproduction, I know, my apologies).

Setting `CARGO_INCREMENTAL=0` fixes the issue for me. Deleting my target directory and rebuilding also fixes the issue.

However, at revision 214ac8d, everything builds fine with `cargo build`. If I then remove 2 lines of dead code and `cargo build`, the panic occurs.

### Meta

`rustc --version --verbose`:
```
PS C:\projects\chitin_stars> rustc --version --verbose
rustc 1.60.0-nightly (1409c015b 2022-01-11)
binary: rustc
commit-hash: 1409c015b44a4d4d38bef2250b2a37c17b8b7463
commit-date: 2022-01-11
host: x86_64-pc-windows-msvc
release: 1.60.0-nightly
LLVM version: 13.0.0

```

### Error output

```
thread 'rustc' panicked at 'index out of bounds: the len is 180 but the index is 180', compiler\rustc_query_impl\src\on_disk_cache.rs:726:18
```

<details><summary><strong>Backtrace</strong></summary>
<p>

Note, the `RUST_BACKTRACE=1` produces an empty backtrace for me, here is the `RUST_BACKTRACE=full`

```
stack backtrace:
   0:     0x7ffdc8df609f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1cbc9437c25bbb3a
   1:     0x7ffdc8e2510a - core::fmt::write::hba4814166b4d0ea8
   2:     0x7ffdc8de7408 - <std::io::IoSliceMut as core::fmt::Debug>::fmt::hb8365da922cb93b3
   3:     0x7ffdc8df9acb - std::panicking::take_hook::ha4f8099bbfce64a0
   4:     0x7ffdc8df95c4 - std::panicking::take_hook::ha4f8099bbfce64a0
   5:     0x7ffda1553c36 - <rustc_middle[9a5348a1b7968334]::ty::SymbolName as core[c599802ae0e8ca0d]::fmt::Debug>::fmt
   6:     0x7ffdc8dfa1f6 - std::panicking::rust_panic_with_hook::h26db5fb7733042a0
   7:     0x7ffdc8df9ffb - <std::panicking::begin_panic_handler::StrPanicPayload as core::panic::BoxMeUp>::get::hc540f1f2d59e426e
   8:     0x7ffdc8df69e7 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1cbc9437c25bbb3a
   9:     0x7ffdc8df9cf9 - rust_begin_unwind
  10:     0x7ffdc8e57a20 - core::panicking::panic_fmt::h5e1675c0cda3a84f
  11:     0x7ffdc8e579e7 - core::panicking::panic_bounds_check::h8a4fb76db5fb5caa
  12:     0x7ffda4df46eb - <rustc_span[5556f15f9cb8485a]::span_encoding::Span as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  13:     0x7ffda4d7cb77 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  14:     0x7ffda4c58959 - rustc_query_impl[c295d1e28666884]::query_callbacks::diagnostic_hir_wf_check
  15:     0x7ffda4d7c5bd - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  16:     0x7ffda4c5862a - rustc_query_impl[c295d1e28666884]::query_callbacks::diagnostic_hir_wf_check
  17:     0x7ffda4bf9493 - <rustc_middle[9a5348a1b7968334]::ty::adt::AdtSizedConstraint as rustc_query_impl[c295d1e28666884]::values::Value>::from_cycle_error
  18:     0x7ffda4be57f4 - <rustc_query_impl[c295d1e28666884]::Queries as rustc_middle[9a5348a1b7968334]::ty::query::QueryEngine>::try_mark_green
  19:     0x7ffda4d6b749 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  20:     0x7ffda4c6eb2c - <rustc_query_impl[c295d1e28666884]::on_disk_cache::OnDiskCache>::load_side_effects
  21:     0x7ffda4d6b838 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  22:     0x7ffda4c1d93a - RNvXsd_NtCs12JCngwkpoK_16rustc_query_impl13on_disk_cacheRINtNtNtNtCs7uRT0406iWa_3std11collections4hash3set7HashSetNtNtCs7kgavpGhC4g_10rustc_span6def_id10LocalDefIdINtNtCsgXOILztEN1p_4core4hash18BuildHasherDefaultNtCs8sYYDNXeHys_10rustc_hash8FxHasherEEINtN
  23:     0x7ffda4c1f203 - RNvXsd_NtCs12JCngwkpoK_16rustc_query_impl13on_disk_cacheRINtNtNtNtCs7uRT0406iWa_3std11collections4hash3set7HashSetNtNtCs7kgavpGhC4g_10rustc_span6def_id10LocalDefIdINtNtCsgXOILztEN1p_4core4hash18BuildHasherDefaultNtCs8sYYDNXeHys_10rustc_hash8FxHasherEEINtN
  24:     0x7ffda4d6b943 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  25:     0x7ffda4be596c - <rustc_query_impl[c295d1e28666884]::Queries as rustc_middle[9a5348a1b7968334]::ty::query::QueryEngine>::try_mark_green
  26:     0x7ffda4d6b749 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  27:     0x7ffda4c272d1 - RNvXsd_NtCs12JCngwkpoK_16rustc_query_impl13on_disk_cacheRINtNtNtNtCs7uRT0406iWa_3std11collections4hash3set7HashSetNtNtCs7kgavpGhC4g_10rustc_span6def_id10LocalDefIdINtNtCsgXOILztEN1p_4core4hash18BuildHasherDefaultNtCs8sYYDNXeHys_10rustc_hash8FxHasherEEINtN
  28:     0x7ffda4c25ff6 - RNvXsd_NtCs12JCngwkpoK_16rustc_query_impl13on_disk_cacheRINtNtNtNtCs7uRT0406iWa_3std11collections4hash3set7HashSetNtNtCs7kgavpGhC4g_10rustc_span6def_id10LocalDefIdINtNtCsgXOILztEN1p_4core4hash18BuildHasherDefaultNtCs8sYYDNXeHys_10rustc_hash8FxHasherEEINtN
  29:     0x7ffda4d75d43 - <&[rustc_ast[ffafe597e25034ed]::ast::InlineAsmTemplatePiece] as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  30:     0x7ffda4c54b32 - rustc_query_impl[c295d1e28666884]::query_callbacks::diagnostic_hir_wf_check
  31:     0x7ffda4c30186 - RNvXsd_NtCs12JCngwkpoK_16rustc_query_impl13on_disk_cacheRINtNtNtNtCs7uRT0406iWa_3std11collections4hash3set7HashSetNtNtCs7kgavpGhC4g_10rustc_span6def_id10LocalDefIdINtNtCsgXOILztEN1p_4core4hash18BuildHasherDefaultNtCs8sYYDNXeHys_10rustc_hash8FxHasherEEINtN
  32:     0x7ffda4c590a3 - rustc_query_impl[c295d1e28666884]::query_callbacks::diagnostic_hir_wf_check
  33:     0x7ffda4c07398 - <rustc_middle[9a5348a1b7968334]::mir::interpret::GlobalId as rustc_query_impl[c295d1e28666884]::keys::Key>::default_span
  34:     0x7ffda4c6cae7 - <rustc_query_impl[c295d1e28666884]::on_disk_cache::OnDiskCache>::load_side_effects
  35:     0x7ffda4de9c89 - <alloc[9d38bed4de9fe7a2]::vec::Vec<u8> as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  36:     0x7ffda4cdb8c9 - <&rustc_index[8d1c4d35d70c6212]::vec::IndexVec<rustc_middle[9a5348a1b7968334]::mir::Promoted, rustc_middle[9a5348a1b7968334]::mir::Body> as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  37:     0x7ffda4ad780b - <rustc_mir_dataflow[47b59de76a88bb6e]::impls::storage_liveness::MaybeRequiresStorage as rustc_mir_dataflow[47b59de76a88bb6e]::framework::AnalysisDomain>::initialize_start_block
  38:     0x7ffda4baf452 - <rustc_query_impl[c295d1e28666884]::Queries>::as_dyn
  39:     0x7ffda4a5ddfd - <rustc_mir_dataflow[47b59de76a88bb6e]::impls::storage_liveness::MaybeRequiresStorage as rustc_mir_dataflow[47b59de76a88bb6e]::framework::AnalysisDomain>::initialize_start_block
  40:     0x7ffda4b18c09 - <rustc_mir_dataflow[47b59de76a88bb6e]::impls::storage_liveness::MaybeRequiresStorage as rustc_mir_dataflow[47b59de76a88bb6e]::framework::AnalysisDomain>::initialize_start_block
  41:     0x7ffda59ec627 - <rustc_middle[9a5348a1b7968334]::ty::context::TyCtxt>::instance_mir
  42:     0x7ffda3ae3794 - <rustc_monomorphize[720d5d561c9598aa]::collector::RootCollector as rustc_hir[8a9015b0b0718f02]::itemlikevisit::ItemLikeVisitor>::visit_impl_item
  43:     0x7ffda3adccde - <rustc_monomorphize[720d5d561c9598aa]::partitioning::MonoItemPlacement as core[c599802ae0e8ca0d]::fmt::Debug>::fmt
  44:     0x7ffda3ac5c6e - <core[c599802ae0e8ca0d]::option::Option<rustc_middle[9a5348a1b7968334]::middle::privacy::AccessLevel> as rustc_privacy[1c188a2c9bf6307e]::VisibilityLike>::new_min
  45:     0x7ffda3adb233 - <rustc_monomorphize[720d5d561c9598aa]::partitioning::MonoItemPlacement as core[c599802ae0e8ca0d]::fmt::Debug>::fmt
  46:     0x7ffda3ad5ed3 - <rustc_monomorphize[720d5d561c9598aa]::partitioning::default::DefaultPartitioning as rustc_monomorphize[720d5d561c9598aa]::partitioning::Partitioner>::internalize_symbols
  47:     0x7ffda4de8a53 - <alloc[9d38bed4de9fe7a2]::vec::Vec<u8> as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  48:     0x7ffda4d1c84a - <&rustc_index[8d1c4d35d70c6212]::vec::IndexVec<rustc_middle[9a5348a1b7968334]::mir::Promoted, rustc_middle[9a5348a1b7968334]::mir::Body> as rustc_serialize[84cbc18e011cc316]::serialize::Decodable<rustc_query_impl[c295d1e28666884]::on_disk_cache::CacheDecoder>>::decode
  49:     0x7ffda4bcab8c - <rustc_query_impl[c295d1e28666884]::Queries>::as_dyn
  50:     0x7ffda4abb07f - <rustc_mir_dataflow[47b59de76a88bb6e]::impls::storage_liveness::MaybeRequiresStorage as rustc_mir_dataflow[47b59de76a88bb6e]::framework::AnalysisDomain>::initialize_start_block
  51:     0x7ffda4b50b8a - <rustc_mir_dataflow[47b59de76a88bb6e]::impls::storage_liveness::MaybeRequiresStorage as rustc_mir_dataflow[47b59de76a88bb6e]::framework::AnalysisDomain>::initialize_start_block
  52:     0x7ffda4be094e - <rustc_query_impl[c295d1e28666884]::Queries as rustc_middle[9a5348a1b7968334]::ty::query::QueryEngine>::try_mark_green
  53:     0x7ffda17957c0 - <rustc_codegen_llvm[5c71da2058f94c44]::LlvmCodegenBackend as rustc_codegen_ssa[23c6f6855f693bee]::traits::backend::CodegenBackend>::codegen_crate
  54:     0x7ffda16f199b - <rustc_interface[dea2a2a67588089]::passes::boxed_resolver::BoxedResolver>::to_resolver_outputs
  55:     0x7ffda16591c9 - <rustc_interface[dea2a2a67588089]::queries::Queries>::ongoing_codegen
  56:     0x7ffda150573f - <rustc_driver[7959e63f47696881]::args::Error as core[c599802ae0e8ca0d]::fmt::Debug>::fmt
  57:     0x7ffda14e2445 - <tracing_subscriber[aafa3d21bab35d8]::fmt::writer::TestWriter as std[57552265efe0c730]::io::Write>::flush
  58:     0x7ffda1503091 - <tracing_subscriber[aafa3d21bab35d8]::fmt::writer::TestWriter as std[57552265efe0c730]::io::Write>::flush
  59:     0x7ffda14e5a17 - <tracing_subscriber[aafa3d21bab35d8]::fmt::writer::TestWriter as std[57552265efe0c730]::io::Write>::flush
  60:     0x7ffda14e465c - <tracing_subscriber[aafa3d21bab35d8]::fmt::writer::TestWriter as std[57552265efe0c730]::io::Write>::flush
  61:     0x7ffda1509008 - <rustc_middle[9a5348a1b7968334]::ty::SymbolName as core[c599802ae0e8ca0d]::fmt::Debug>::fmt
  62:     0x7ffdc8e09bac - std::sys::windows::thread::Thread::new::hea3b39ea0f700098
  63:     0x7ffe0f637034 - BaseThreadInitThunk
  64:     0x7ffe11482651 - RtlUserThreadStart
```

</p>
</details>

**Comment by cbeuw** [comment]
Should have been fixed by #92533. Try the most recent nightly?

**Comment by SwampWalker** [comment]
Confirmed, no longer an issue.
